trustmanager/keyfilestore.go

```diff
@@ -106,64 +109,120 @@ func (s *KeyMemoryStore) ListKeys() []string {
 	return listKeys(s)
 }
 
-func addKey(s LimitedFileStore, name string, privKey data.PrivateKey) error {
+// RemoveKey removes the key from the keystore
+func (s *KeyMemoryStore) RemoveKey(name string) error {
+	return removeKey(s, name)
+}
+
+func addKey(s LimitedFileStore, passphraseRetriever PassphraseRetriever, name, alias string, privKey data.PrivateKey) error {
 	pemPrivKey, err := KeyToPEM(privKey)
 	if err != nil {
 		return err
 	}
 
-	return s.Add(name, pemPrivKey)
-}
-
-func getKey(s LimitedFileStore, name string) (data.PrivateKey, error) {
-	keyBytes, err := s.Get(name)
-	if err != nil {
-		return nil, err
+	attempts := 0
+	passphrase := ""
+	giveup := false
+	for {
+		passphrase, giveup, err = passphraseRetriever(name, alias, true, attempts)
+		if err != nil {
+			attempts++
+			continue
+		}
+		if giveup {
+			return errors.New("obtaining passphrase failed")
+		}
+		if attempts > 10 {
+			return errors.New("maximum number of passphrase attempts exceeded")
+		}
+		break
 	}
 
-	// Convert PEM encoded bytes back to a PrivateKey
-	privKey, err := ParsePEMPrivateKey(keyBytes, "")
-	if err != nil {
-		return nil, err
+	if passphrase != "" {
+		pemPrivKey, err = EncryptPrivateKey(privKey, passphrase)
+		if err != nil {
+			return err
+		}
 	}
 
-	return privKey, nil
+	return s.Add(name+"_"+alias, pemPrivKey)
 }
 
-func addEncryptedKey(s LimitedFileStore, name string, privKey data.PrivateKey, passphrase string) error {
-	encryptedPrivKey, err := EncryptPrivateKey(privKey, passphrase)
-	if err != nil {
-		return err
+func getKeyAlias(s LimitedFileStore, keyID string) (string, error) {
+	files := s.ListFiles(true)
+	name := strings.TrimSpace(strings.TrimSuffix(filepath.Base(keyID), filepath.Ext(keyID)))
+
+	for _, file := range files {
+		lastPathSeparator := strings.LastIndexAny(file, "/\\")
+		filename := file[lastPathSeparator+1:]
+
+		if strings.HasPrefix(filename, name) {
+			aliasPlusDotKey := strings.TrimPrefix(filename, name+"_")
+			retVal := strings.TrimSuffix(aliasPlusDotKey, "."+keyExtension)
+			return retVal, nil
+		}
 	}
```

Comment:
I'm not crazy about this way of doing things. It means some operations like removing every key will become O(n^2). It's probably not a practical concern since we shouldn't have many keys at a time.

IMHO it would be a little better to have a subdirectory for each key, so we could list just that subdirectory. But I don't think it's worth spending time changing it at this point.